### ISSUE TITLE
knew raw create view has string constant with an E, db Table Definition missing E

### ISSUE THREAD

**Issue by kidCaulfield** [knex/knex#4850]
# Environment

Knex version:
Database + version:Postgres + 13.2
OS: MacOS 10.15.2

# Bug

I am trying to create a view with `regex_replace` that will take an array nested in a jsonb column and convert it to Postgres array format which will allow me to later use the Postgres `ANY()` function on this column later.

eg.

```js
knex.raw(`
  CREATE VIEW colorado
  AS SELECT
    regexp_replace(application.colorado ->> 'type_checkbox', E'\\[(([^][]|\\[([^][]|\\[[^][]*\\])*\\])*)\\]', E'{\\1}', 'g')::varchar[] AS type_checkbox
  FROM application
`)
```

However after migrating this into the database, the Table Definition is missing the E on the string constant.

eg. Table Definition:
```js
SELECT
    regexp_replace(application.colorado ->> 'type_checkbox'::text, '\\[(([^][]|\\[([^][]|\\[[^][]*\\])*\\])*)\\]', '{\\1}'::text, 'g'::text)::varchar[] AS type_checkbox
FROM application
```

Which causes an error:

`ERROR: invalid regular expression: parentheses () not balanced` when querying the view.

Postgres documentation mentions here https://www.postgresql.org/docs/9.0/sql-syntax-lexical.html :

If you need to use a backslash escape to represent a special character, write the string constant with an E to be sure it will be handled the same way in future releases.

I need knex make a view have a string constant with an E that will stick in the Table Definition.

**Comment by OlivierCavadenti** [comment]
Just in case, you can now createView in Knex (0.95.14) so you can just make raw in 'select' function : https://knexjs.org/#Schema-createView
like, for the "as" query in view :  select(knew.raw(.....)).from('application').
Now for you problem I will check later and make tests.

**Comment by mercmobily** [comment]
Hi! Thanks for the detailed example and for citing the PostgreSQL docs.

TL;DR:

**I think it would be worthwhile to add a small doc note under the PostgreSQL `raw` section to call out deparser behaviour around E-strings and suggest dollar-quoting for complex regexes.**

This looks like PostgreSQL normalising the stored view definition, not Knex stripping the `E` prefix. Knex passes the raw string as-is; `psql` and `pg_catalog.pg_get_viewdef` often rewrite escapes in the canonical definition.

When you send a raw with `E'...'`:

```js
knex.raw(`CREATE VIEW ... regexp_replace(..., E'\\\\[ ... \\]', E'{\\\\1}', 'g')::varchar[] ...`)
```

Knex does not alter that SQL. The behaviour you observe appears after the view is stored and you look at the “table definition” via the driver, which uses PostgreSQL’s canonical deparser; with `standard_conforming_strings` on (default on modern PG), the deparser typically emits standard strings without `E`, doubling backslashes to preserve semantics. That’s why you’re seeing `'\\\\[ ... ]'` without `E` — it’s the same pattern encoded using standard strings, not a change introduced by Knex.

The regex error suggests that in your case the deparsed form did not preserve the escapes as intended. Two practical mitigations you can try that avoid depending on E-strings in the stored view:
- Use dollar-quoted strings for the regex: `$$...$$` does not treat backslashes specially and avoids the E/non‑E differences entirely.
- Ensure `standard_conforming_strings` is on (it usually is since PG 9.1), and adjust the literal accordingly so the deparsed form remains equivalent.

Separately, since `0.95.14` Knex supports `schema.createView`, which lets you compose the view and still drop in `raw` for complex expressions. That won’t change the deparser behaviour, but it can make the definition more maintainable.

Technically: there is no code path in Knex that rewrites `E'...'`; it’s passed through untouched.
